app/controllers/result_filters/location_controller.rb

```diff
@@ -2,21 +2,70 @@ module ResultFilters
   class LocationController < ApplicationController
     include FilterParameters
 
+    before_action :create_view
+
     def new; end
 
     def create
-      if(!params[:l])
-        flash[:error] = "Please choose an option"
-        redirect_to location_path(params_without_unsupported_location_params)
+      build_params = strip(filter_params.clone)
+
+      case @view.selected_option
+      when @view.no_option
+        handle_no_option(build_params)
+      when @view.location_option
+        handle_location_option(build_params)
       else
-        redirect_to results_path(params_without_unsupported_location_params)
+        redirect_to results_path(build_params)
       end
     end
 
-  private
+    private
+
+    def handle_no_option(build_params)
+      flash[:error] = "Please choose an option"
+      redirect_to location_path(build_params)
+    end
+
+    def handle_location_option(build_params)
+      geocode_params = geocode_params_for(@view.location_query)
+      if !geocode_params
+        build_geocode_errors
+        redirect_to location_path(build_params)
+      else
+        merged_params = build_params.merge(geocode_params)
+        redirect_to results_path(merged_params)
+      end
+    end
 
-    def params_without_unsupported_location_params
-      filter_params.except(:lat, :lng, :rad, :query, :loc, :lq)
+    def build_geocode_errors
+      flash[:error] = "Postcode, town or city"
+      flash[:unknown_location] = "We couldn't find this location, please check your input and try again."
+      flash[:query] = @view.location_query
+    end
+
+    def create_view
+      @view = ResultFilters::LocationView.new(params: filter_params)
+    end
+
+    def strip(params)
+      params.reject { |_, v| v == "" }
+    end
+
+    def geocode_params_for(query)
+      # Google's Geocoding API is not always reliable
+      # https://stackoverflow.com/questions/2647086/googles-geocoder-returns-wrong-country-ignoring-the-region-hint
+      # e.g. A query 'FOOBAR' still gets a hit from the Netherlands even if
+      # the region is restricted by 'uk'.
+      # Appending 'uk' to the query appears to prevent this from happening.
```

Comment:
Or an alternative approach would be to *not* append 'uk' and instead select / filter only those hits that are from the 'uk'. Anyone have a preference?

Reply:
In fact, we should only be dealing with England, I thought. Should we limit it to that, even?

Reply:
Maybe filtering the results (or maybe both) would protect us more if we get strange results. There's no guarantee that adding UK to the query will always return the results we want. That said, it's highly unlikely I suppose and if it adds a lot of complexity then 🤷‍♂

Reply:
OK no worries - found a solution that restricts results to England only. Just requires tweaking the params.

```
irb(main):041:0> Geocoder.search('Edinburgh', components: 'country:england')
=> []
irb(main):042:0> Geocoder.search('FOOBAR', components: 'country:england')
=> []
```

Reply:
Looks like a tweak of the params fixes the issue

```
 Geocoder.search('FOOBAR', components: 'country:UK')
=> [#<Geocoder::Result::Google:0x00007fb41f8e1068 @data={"address_components"=>[{"long_name"=>"1", "short_name"=>"1", "types"=>["street_number"]}, {"long_name"=>"Rathbone Square", "short_name"=>"Rathbone Square", "types"=>["route"]}, {"long_name"=>"Fitzrovia", "short_name"=>"Fitzrovia", "types"=>["neighborhood", "political"]}, {"long_name"=>"London", "short_name"=>"London", "types"=>["postal_town"]}, {"long_name"=>"Greater London", "short_name"=>"Greater London", "types"=>["administrative_area_level_2", "political"]}, {"long_name"=>"England", "short_name"=>"England", "types"=>["administrative_area_level_1", "political"]}, {"long_name"=>"United Kingdom", "short_name"=>"GB", "types"=>["country", "political"]}, {"long_name"=>"W1T 1FB", "short_name"=>"W1T 1FB", "types"=>["postal_code"]}], "formatted_address"=>"1 Rathbone Square, Fitzrovia, London W1T 1FB, UK", "geometry"=>{"location"=>{"lat"=>51.5167849, "lng"=>-0.1340803}, "location_type"=>"ROOFTOP", "viewport"=>{"northeast"=>{"lat"=>51.5181338802915, "lng"=>-0.132731319708498}, "southwest"=>{"lat"=>51.5154359197085, "lng"=>-0.135429280291502}}}, "place_id"=>"ChIJUXyR96obdkgRUs5NVnKio00", "plus_code"=>{"compound_code"=>"GV88+P9 London, United Kingdom", "global_code"=>"9C3XGV88+P9"}, "types"=>["bar", "establishment", "point_of_interest"]}, @cache_hit=nil>]
```

Reply:
Doesn't look like the search can be further refined to England though